### PR TITLE
[REST] add upload service

### PR DESCRIPTION
**Pull request by fkroockmann**
- Add upload resource service (Move a file into a Tmp directory of BackBee)
- Add listener onFlush of a content revision (Move the file into a Media directory)

### CODE REVIEW

File: Config/events.yml [hunk 1]
```diff
@@ -66,7 +66,11 @@ site.layout.postremove:
 revision.postload:
     listeners:
         - [BackBee\Event\Listener\RevisionListener, onPostLoad]
-
+        
+revision.onflush:
+    listeners:
+        - [BackBee\Event\Listener\RevisionListener, onFlushElementFile]
```
Comment (eric-chau) [inline Config/events.yml]: Move this listener to ``BackBee\ClassContent\Listener`` please.

File: Event/Listener/RevisionListener.php [hunk 2]
```diff
@@ -107,6 +112,47 @@ public static function onFlushContent(Event $event)
             }
         }
     }
+    
+    public static function onFlushElementFile(Event $event) 
+    {
+        $dispatcher = $event->getDispatcher();
+        $application = $dispatcher->getApplication();
```
Comment (eric-chau) [inline Event/Listener/RevisionListener.php]: ``$application = $event->getApplication()``.

File: Event/Listener/RevisionListener.php [hunk 3]
```diff
@@ -107,6 +112,47 @@ public static function onFlushContent(Event $event)
...
+        $content = $revision->getContent();
+        
+        if ($content instanceof ElementFile) {
+            $moveFrom = $content->path;
+            
+            $content->originalname = Utils\String::toPath($content->originalname);
+            $content->path = Media::getPathFromContent($content);
+            
+            $moveTo = $content->path;
+
```
Comment (eric-chau) [inline Event/Listener/RevisionListener.php]: remove one line.

File: Event/Listener/RevisionListener.php [hunk 4]
```diff
@@ -107,6 +112,47 @@ public static function onFlushContent(Event $event)
...
+        if ($content instanceof ElementFile) {
+            $moveFrom = $content->path;
+            
+            $content->originalname = Utils\String::toPath($content->originalname);
+            $content->path = Media::getPathFromContent($content);
+            
+            $moveTo = $content->path;
+            
+            
+            Utils\File\File::resolveFilepath($moveTo, NULL, array('base_dir' => $application->getMediaDir()));
```
Comment (eric-chau) [inline Event/Listener/RevisionListener.php]: Write null in lowercase please.

File: Rest/Controller/ResourceController.php [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        $data = [];
+        
+        if ($files->count() === 1) {
+            foreach ($files as $file) {
+                if (null !== $file) {
+                    if ($file->isValid()) {
+                        if ($file->getClientSize() <= $file->getMaxFilesize()) {
+                            $data['originalname'] = $file->getClientOriginalName();
+                            
+                            if (null !== $uploadConfig && isset($uploadConfig['base_64']) && $uploadConfig['base_64'] === true) {
```
Comment (eric-chau) [inline Rest/Controller/ResourceController.php]: Remove base64 rules cause not managed at moment.

File: Rest/Controller/ResourceController.php [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+                                $data['path'] = $file->getRealPath() . '.' . $file->guessExtension();
+                                $data['src'] = base64_encode(file_get_contents($file->getRealPath()));
+                            } else {
+                                $fileName = md5($file->getClientOriginalName() . time() . $file->getClientSize()) . '.' . $file->guessExtension();
+                                $data['path'] = $tmpDirectory . DIRECTORY_SEPARATOR . $fileName;
+                                
+                                $file->move($tmpDirectory, $fileName);
+                            }
+                        } else {
+                            throw new BadRequestHttpException('Too big file');
```
Comment (eric-chau) [inline Rest/Controller/ResourceController.php]: Add more informations into your exception message please (like the max file size supported, etc.)

File: Rest/Controller/ResourceController.php [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+                    }
+                }
+
+                break;
+            }
+        } else {
+            if ($files->count() === 0) {
+                throw new NotFoundHttpException('No file to upload');
+            } else {
+                throw new BadRequestHttpException('Too many file to upload');
```
Comment (eric-chau) [inline Rest/Controller/ResourceController.php]: Be more explicit in your message, something like ``You can upload only one file by request.``

File: Config/events.yml (hunk 1 above)
Comment (fkroockmann) [inline Config/events.yml]: This file already exist in this folder, i just add a onFlushElementFile function

**Comment by fkroockmann** [comment]
Thanks for review @eric-chau.

**Comment by eric-chau** [comment]
:+1: good job, thank you @fkroockmann.

**Pull request merged**